subsystems/DrivetrainBase.java

```diff
@@ -19,33 +26,99 @@
+    // create a hashtable to look up constants
+    protected DriveConstants constants;
+
+    // create variables needed for odometry.
     protected AHRS ahrs;
     protected FPID turnFPID;
     protected double turnFPIDThrottle;
     protected boolean invertGyro;
+    protected DifferentialDriveOdometry odometry;
+    private final LimitedPoseMap poseHistory;
+    private double lastAngleSetpoint = Double.NaN;
+    private final PIDController turnController;
+
+    public static class DriveConstants {
+        public final FPID turnFPID;
+        public final double turnPIDThrottle, turnPIDToleranceDeg, wheelDiameterMeters;
+        public final int historyLimit;
+        public final boolean invertGyro;
 
-    // placeholders for a SparkMax-based drivetrain
-    /*
-    protected CANSparkMax leftLeader = new CANSparkMax(0, MotorType.kBrushless);
-    protected CANSparkMax rightLeader = new CANSparkMax(1, MotorType.kBrushless);
-    protected CANSparkMax leftFollower = new CANSparkMax(2, MotorType.kBrushless);
-    protected CANSparkMax rightFollower = new CANSparkMax(3, MotorType.kBrushless);
-    */
+        public DriveConstants(FPID turnFPID, double turnPIDThrottle, double turnPIDToleranceDeg, double wheelDiameterMeters, int historyLimit, boolean invertGyro) {
+            this.turnFPID = turnFPID;
+            this.turnPIDThrottle = turnPIDThrottle;
+            this.turnPIDToleranceDeg = turnPIDToleranceDeg;
+            this.wheelDiameterMeters = wheelDiameterMeters;
+            this.historyLimit = historyLimit;
+            this.invertGyro = invertGyro;
+        }
+    }
 
     // set all of the variables from the subclass to this abstract class
-    public DrivetrainBase(WPI_TalonFX leftLeader, WPI_TalonFX rightLeader, WPI_TalonFX leftFollower, WPI_TalonFX rightFollower, FPID turnFPID, double turnFPIDThrottle, boolean invertGyro) {
-        super(new PIDController(turnFPID.kP, turnFPID.kI, turnFPID.kD));
-        
-        this.leftLeader = leftLeader;
-        this.rightLeader = rightLeader;
-        this.leftFollower = leftFollower;
-        this.rightFollower = rightFollower;
-        this.turnFPID = turnFPID;
-        this.turnFPIDThrottle = turnFPIDThrottle;
-        this.invertGyro = invertGyro;
-        leftGroup = new SpeedControllerGroup(leftLeader, leftFollower);
-        rightGroup = new SpeedControllerGroup(rightLeader, rightFollower);
+    // MotorIDs: a arrays of integer CAN IDs used to make motors. index 0 should ALWAYS be the leader motor, and anything else is a follower.
+    public DrivetrainBase(DriveConstants constants, int[] leftMotorIDs, int[] rightMotorIDs) {
+        super(new PIDController(constants.turnFPID.kP, constants.turnFPID.kI, constants.turnFPID.kD));        
+        // disable PID control on start
+        this.disable();
+        this.constants = constants;
+        // int i = 0;
+        // make arrays to put each type of motor in
+        WPI_TalonFX[] leaders = new WPI_TalonFX[2];
```

Comment:
You should have separate left leader and right leader variables, you don't need an array because you just unpack it later. The array just adds confusion.

Reply:
Oh yeah lol